### ISSUE TITLE
Validation loop is extremely slow compared to training loop.

### ISSUE THREAD

**Issue by zeeshansayyed** [allenai/allennlp#3909]
I am training `seq2seq` models. I have tried both `simple_seq2seq` and `composed_seq2seq` models. I used `seq2seq` dataset reader for loading both the training and the validation data. The model seems to work well so far. (At least the loss decreases) for the first 50 epochs. I am facing one big problem. The validation loop is very very slow compared to the training loop. For a batch size of 16, I am getting 28it/s in training whereas I get 4s/it in validation. I checked GPU-Util during validation loop and it is in the 90s. What do you think might be going wrong?

Note: I am using P100 to train

My iterator and trainer look as follows:

```jsonnet
  iterator: {
    type: 'bucket',
    batch_size: 16,
    sorting_keys: [['source_tokens', 'num_tokens']],
  },

  trainer: {
    num_epochs: 100,
    optimizer: {
      type: 'adam',
      lr: 0.01,
    },
    patience: 5,
    cuda_device: 0,
  },
```

**Comment by DeNeutoy** [comment]
@zeeshansayyed please include your model config. 

You might experiment with these two parameters:
https://github.com/allenai/allennlp/blob/master/allennlp/models/encoder_decoders/simple_seq2seq.py#L52

https://github.com/allenai/allennlp/blob/master/allennlp/models/encoder_decoders/simple_seq2seq.py#L62

It's possible that the BLEU computation is making your decode loop very slow. Also, if you've set a very high beam size, that will also slow things down too. Those are the ideas I have off the top of my head, hopefully helpful!

**Comment by zeeshansayyed** [comment]
Thanks @DeNeutoy I will try turning bleu off and reducing the beam size and will report back what the effect was. 

Here's my mode config:

```jsonnet
local EMBEDDING_DIM = 16;
local NUM_HEADS = 2;
model: {
    type: 'composed_seq2seq',
    source_text_embedder: {
      type: 'basic',
      token_embedders: {
        tokens: {
          type: 'embedding',
          embedding_dim: EMBEDDING_DIM,
        },
      },
    },
    encoder: {
      type: 'stacked_self_attention',
      input_dim: EMBEDDING_DIM,
      hidden_dim: EMBEDDING_DIM,
      projection_dim: EMBEDDING_DIM,
      feedforward_hidden_dim: EMBEDDING_DIM,
      num_layers: 2,
      num_attention_heads: NUM_HEADS,
    },
    decoder: {
      target_namespace: 'tokens',
      target_embedder: {
        vocab_namespace: 'tokens',
        embedding_dim: EMBEDDING_DIM,
      },
      decoder_net: {
        type: 'stacked_self_attention',
        target_embedding_dim: EMBEDDING_DIM,
        decoding_dim: EMBEDDING_DIM,
        feedforward_hidden_dim: EMBEDDING_DIM,
        num_layers: 2,
        num_attention_heads: 2,
      },
      tie_output_embedding: true,
      max_decoding_steps: 500,
      beam_size: 5,
      label_smoothing_ratio: 0.1,
    },
    tied_source_embedder_key: 'tokens',
  },
```

**Comment by bryant1410** [comment]
Have you tried with allennlp from the master branch?

**Comment by zeeshansayyed** [comment]
@bryant1410 No I haven't. But I think I have narrowed it down to this: Beam Search gets applied even on the validation set. This is exacerbated due to some error, which is causing my model to not predict the EOS token and keeps on predicting sequences of length `max_decoding_steps`. I tested the dataset outside of AllenNLP (in Fairseq) and it works well there. I am assuming that it could because of how CharacterTokenizer is handling the tokenization, though I am not sure. I will try tokenizing it beforehand and using a space tokenizer to see if the problem persists.

**Comment by tshu-w** [comment]
I had a similar problem, beam search was especially slow. Is there any way to improve it, I'm not very familiar with generation.

**Comment by epwalsh** [comment]
One simple thing you can do is increase the batch size in validation data loader. You could probably get away with at least twice the batch size compared to train.

**Comment by bryant1410** [comment]
Note however in beam search you already have a larger batch size than in training.

**Comment by epwalsh** [comment]
That's true, but even so I've been able to run at least 2x the train batch size in validation with the CopyNet model.

**Comment by JohnGiorgi** [comment]
I know this thread is old but, I am also seeing really slow validation times. I am using a model based on CopyNet, and validation is 2-3X slower than training. This is true even despite my best efforts to speed things up:

- using the latest version of allennlp/allennlp-models
- using a validation dataloader that sorts based on source text length, with a batch size 32X (!) larger than training and `padding_noise` set to 0
- using greedy decoding by setting `beam_size` to 1
- using a reasonable `max_steps` in `beam_search` (i.e. 128)

Is there anything else that is known to slow down validation? It seems extreme that it would be 2-3X longer than training.

**Comment by epwalsh** [comment]
`CopyNet` has always been slow with validation, mainly because the forward pass during validation is a lot more complex than the training forward pass. But it might be worth trying to profile it. Maybe there's a few bottlenecks that could be improved.

**Comment by JohnGiorgi** [comment]
Gotcha, thanks @epwalsh!
